application/helpers/common_helper.php

```diff
@@ -3737,18 +3737,16 @@ function fixLanguageConsistency($sid, $availlangs = '', $baselang = '')
                 Yii::app()->db->createCommand()->insert('{{group_l10ns}}', $data);
             }
         }
-        reset($langs);
     }
 
-    // @todo Too many queries if survey has more than one language
-    $query = "SELECT * FROM {{questions}} q JOIN {{question_l10ns}} ls ON ls.qid=q.qid WHERE sid='{$sid}'";
-    $result = Yii::app()->db->createCommand($query)->query()->readAll();
+    $query = "SELECT q.qid, ls.question, ls.help FROM {{questions}} q JOIN {{question_l10ns}} ls ON ls.qid=q.qid WHERE sid={$sid} AND language='{$baselang}'";
+    $result = Yii::app()->db->createCommand($query)->query();
     if (count($result) > 0) {
-        foreach ($result as $question) {
-            foreach ($langs as $lang) {
-                $query = "SELECT count(qid) FROM {{question_l10ns}} WHERE qid='{$question['qid']}' AND language='{$lang}'";
-                $gresult = Yii::app()->db->createCommand($query)->queryScalar();
-                if ($gresult < 1) {
+        foreach ($languagesToCheck as $lang) {
+            $query = "SELECT q.qid FROM {{questions}} q JOIN {{question_l10ns}} ls ON ls.qid=q.qid WHERE sid={$sid} AND language='{$lang}'";
```

Comment:
If you would like to minimize queries even more, maybe the where clause could be something like `language in (...)`

Reply:
Unsure on this point : if you have 100 questions with 20 language with Chinese and Arabic and long sentence : it can take a bunch of memory and broken a lot of instance.

I think it(rs best to separate each request by language

Reply:
Yeah, I thought about this too, but I think the number of queries reduced to the number of languages is fine here.
You will rarely see a survey with 50 languages where this would really start to matter.
For that case we could maybe introduce an arbitrary limit on number of languages per survey? 
I could vet our customer base and see what is the current maximum.

Reply:
I see maximum of 23 language. But i see a lot of surveys with more than 300 questions ;)

Reply:
> You will rarely see a survey with 50 languages where this would really start to matter.

yes but increasing time is cheaper than adding memory (i already increase timeout before each update process to allow check integrity before update)

Reply:
I wouldn't limit the number of languages, at least not in the Community Version.
On the cloud version could be something limited by a plan :)

I have seen customers using many (maybe +50) languages.
If you limit the number of languages (at any number of languages), how will they work their surveys?

> Unsure on this point : if you have 100 questions with 20 language with Chinese and Arabic and long sentence : it can take a bunch of memory and broken a lot of instance.

What do you mean by "long sentences"? Only qid is selected.
100 questions x 20 languages = 2000 rows of a single integer. Doens't sound like a lot.
Am I missing something?

Reply:
> What do you mean by "long sentences"? Only qid is selected.
> 100 questions x 20 languages = 2000 rows of a single integer. Doens't sound like a lot.

You're right !